### DISCUSSION TITLE
How to use GreaterEqualThreshold with Real variable

### DISCUSSION THREAD

**Discussion by Arinomo** [OpenModelica/OpenModelica#8600] [Q&A] [answered]
I have a following test models

```.mo
model CompareTest
  parameter Real Max = 10.13;
  Modelica.Blocks.Sources.CombiTimeTable combiTimeTable(extrapolation = Modelica.Blocks.Types.Extrapolation.HoldLastPoint, smoothness = Modelica.Blocks.Types.Smoothness.ConstantSegments, table = [0, 0; 0.5, Max; 1.5, 0]);
  Boolean Flag;
initial equation
  Flag = false;
equation
  when combiTimeTable.y[1] >= Max then
    Flag = true;
  end when;

end CompareTest;

model CompareTest2
 parameter Real Max = 10.13;
  Modelica.Blocks.Sources.CombiTimeTable combiTimeTable(extrapolation = Modelica.Blocks.Types.Extrapolation.HoldLastPoint, smoothness = Modelica.Blocks.Types.Smoothness.ConstantSegments, table = [0, 0; 0.5, Max; 1.5, 0]);
  Modelica.Blocks.Logical.GreaterEqualThreshold greaterEqualThreshold(threshold=Max);
  Boolean Flag;
initial equation
  Flag = false;
equation
  connect(combiTimeTable.y[1], greaterEqualThreshold.u);
  when greaterEqualThreshold.y then
    Flag = true;
  end when;

end CompareTest2;
```
and the variable `Flag` stays `false` on both model.  If i undertand it correctly, it happend because i try to check if two `Real` values are equal. My question is, how can one use GreaterEqualThreshold with a `Real` input and still be sure that the result would be `true` when `u == threshold` is?

**Comment by casella** [comment]
@Arinomo, IMHO there's no way you can get these models to work reliably. The output of greaterEqualThreshold is never going to exceed Max, so y >= Max cannot generate an event reliably.

You really can't do this stuff with Real numbers. If you describe what is your actual use case, I can try to come up with alternative formulations that are guaranteed to work.

**Comment by Arinomo** [comment]
@casella  thank you for your reply.

I'm trying to make a model that change a count-down counter with a different start time based on the input. the idea is to reset the coutner starting point when a certain threshold is exceeded as shown below.

```.mo
package comparePackage
  model Counter
    input Real u;
    constant Real t[3] = {100, 10, 1};
    constant Real x[3] = {8, 80, 800};
    Real counter(start = 1000);
  algorithm
    for i in 1:3 loop
      if u < x[1] then
        break;
      elseif u >= x[i] then
        counter := t[i];
      end if;
    end for;
  end Counter;
  
  model test
    comparePackage.Counter counter annotation(
      Placement(visible = true, transformation(origin = {-8, -2}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  equation
    counter.u = if time > 0.5 then 80 else 5;
  end test;
end comparePackage;
```
so i want to make sure that counter is reset to `10` when `80<=u<800`

**Reply by casella** [reply]
I'm sorry but I don't understand this model, `u `is both a scalar and an array. Additionally, it is undetermined. Can you provide an executable MWE?

**Reply by Arinomo** [reply]
@casella i'm sorry, i made a mistake when typing the code. above is the updated version.

**Comment by casella** [comment, marked as answer]
OK, what I mean is that a condition such as `u >= 80` is only reliable if u actually goes above that value, over time. If it just stays exactly at `u = 80`, it won't be numerically reliable, for the same reason a condition like `u == 80` isn't.

**Reply by Arinomo** [reply]
Thank you @casella for the answer... maybe i'll also try using [compareReal](https://github.com/modelica/Modelica-Compliance/blob/master/ModelicaCompliance/Util.mo) together with greaterThanThrashold and see if it applicable to my use case.

**Reply by casella** [reply]
That doesn't work well either, because the body of functions does not generate events.
